README.md
# KM200 Java API

This is a Java API for heaters with a KM200 gateway.
Heaters of theses brands may use this gateway (according to `/system/brand`):
Bosch, Junkers, Buderus, Nefit, Sieger, Tata, Dakon, Elm, Boulter, Vulcano, Worcester, British Gas, IVT, Geminox, Neckar, Zeus, Milton

## Maven

This package is available in Maven central:
```xml
<dependency>
    <groupId>de.malkusch.km200</groupId>
    <artifactId>km200</artifactId>
    <version>2.0.0</version>
</dependency>
```

## Configuration

To use the API you'll need the uri, gateway password, private password and the salt.

### Private password

The private password is the one you configure yourself when you connect via the
[app](https://play.google.com/store/apps/details?id=com.bosch.tt.buderus) to your heater.
If you forgot your password you can start the "reset internet password" flow in the menu
of your heater and then reassign a new passwort in the app.

### Gateway password

The gateway password is constant and needs to be read out from the menu of your heater (information/internet/login data).

### Salt

I didn't include the salt, because I remember slightly reading somewhere else that Bosch might step in
legally to prevent the publication. So you will have to find the salt on your own, either by searching the internet
or decompiling the [app](https://play.google.com/store/apps/details?id=com.bosch.tt.buderus).
The format is the hexadecimal representation (e.g. `12a0b2…`). Here's an incomplete list of projects in the internet
where you could find the salt:

- [ioBroker.km200](https://github.com/frankjoke/ioBroker.km200/blob/6c0963d671b50cb73f378049448a42cf22a8fecf/km200.js#L13-L17)
- [bosch-thermostat-http-client-python](https://github.com/moustic999/bosch-thermostat-http-client-python/blob/53b2469988c7b25688501669df0981f03a2cbcfa/bosch_thermostat_http/const.py#L5)
- [IPSymconBuderusKM200](https://github.com/demel42/IPSymconBuderusKM200/blob/a71ecedccf8781b607d47692e6c6ebc22a9d1aa3/BuderusKM200/module.php#L683-L686)
- [km200exporter](https://github.com/dirklausen/km200exporter/blob/976344b8f1bec476f25ca1e5619faff12fdccd1d/km200exporter.py#L20)

Additionally I found [this project](https://github.com/bosch-thermostat/bosch-thermostat-client-python/tree/b82b6c46468a647ddf1a2cada38146db8e5ff14f/bosch_thermostat_client/const) which seems to use different salts, depending on the heater's brand.

## Usage

With this library you can access well known endpoints of your KM200 (e.g. `/gateway/DateTime` for your heater's time).
The list of endpoints varies between installations, therefore you have to explore your KM200 yourself. Some endpoints
are writeable (e.g. `/gateway/DateTime`) which you can update with this library as well.

### Examples

Setup a KM200 instance:

```java
var uri = "http://192.168.0.44";
var gatewayPassword = "1234-5678-90ab-cdef";
var privatePassword = "secretExample";
var timeout = Duration.ofSeconds(5);
var salt = "1234567890aabbccddeeff11223344556677889900aabbccddeeffa0a1a2b2d3";

var km200 = new KM200(uri, timeout, gatewayPassword, privatePassword, salt);
```

Read and update the heater's time:

```java
// Read the heater's time
var time = km200.queryString("/gateway/DateTime");
System.out.println(time);

// Update the heater's time
km200.update("/gateway/DateTime", LocalDateTime.now());
```

Explore all endpoints:

```java
km200.endpoints().forEach(System.out::println);
```

### Thread safety

Code wise the API is thread safe, however your KM200 itself might not be. I did observe issues when querying my KM200 concurrently. It performed badly and with errors. It is advised to use this API not concurrently.

## License

The encryption and decryption was extracted from the [OpenHAB](https://github.com/openhab/openhab1-addons/tree/v1.10.0/bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal) project which itself is under the 
Eclipse Public License 2.0. For simplicity I chose to use the same license for this project.


CHANGELOG.md
# Version: 2.0.6

* [#1](https://github.com/malkusch/km200/pull/1): Bump failsafe from 3.0.0 to 3.2.4
* [#2](https://github.com/malkusch/km200/pull/2): Bump jackson-databind from 2.13.0 to 2.13.4
* [#3](https://github.com/malkusch/km200/pull/3): Bump wiremock-jre8 from 2.31.0 to 2.33.2
* [#4](https://github.com/malkusch/km200/pull/4): Bump maven-surefire-plugin from 3.0.0-M5 to 3.0.0-M7
* [#5](https://github.com/malkusch/km200/pull/5): Bump junit-jupiter from 5.8.1 to 5.9.0
* [#6](https://github.com/malkusch/km200/pull/6): Bump wiremock-jre8 from 2.33.2 to 2.34.0
* [#7](https://github.com/malkusch/km200/pull/7): Bump failsafe from 3.2.4 to 3.3.0
* [#8](https://github.com/malkusch/km200/pull/8): Bump junit-jupiter from 5.9.0 to 5.9.1
* [#9](https://github.com/malkusch/km200/pull/9): Bump jackson-databind from 2.13.4 to 2.13.4.1
* [#10](https://github.com/malkusch/km200/pull/10): Bump jackson-databind from 2.13.4.1 to 2.13.4.2
* [#11](https://github.com/malkusch/km200/pull/11): Bump wiremock-jre8 from 2.34.0 to 2.35.0
* [#12](https://github.com/malkusch/km200/pull/12): Bump jackson-databind from 2.13.4.2 to 2.14.0
* [#13](https://github.com/malkusch/km200/pull/13): Bump jackson-databind from 2.14.0 to 2.14.1
* [#14](https://github.com/malkusch/km200/pull/14): Bump oss-parent from 3.0.1 to 3.0.2
* [#15](https://github.com/malkusch/km200/pull/15): Bump maven-surefire-plugin from 3.0.0-M7 to 3.0.0-M8


src/main/java/de/malkusch/km200/KM200.java
package de.malkusch.km200;

import static java.net.http.HttpClient.newBuilder;
import static java.net.http.HttpClient.Redirect.ALWAYS;
import static java.util.Objects.requireNonNull;

import java.io.IOException;
import java.math.BigDecimal;
import java.net.CookieManager;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpRequest.BodyPublishers;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.stream.Stream;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import dev.failsafe.Failsafe;
import dev.failsafe.FailsafeException;
import dev.failsafe.FailsafeExecutor;
import dev.failsafe.RetryPolicy;

/**
 * This is an API for Bosch/Buderus/Junkers heaters with a KM200 gateway.
 * 
 * Although code wise this class is thread safe, it is highly recommended to not
 * use it concurrently. Chances are that your KM200 gateway itself is not thread
 * safe.
 */
public final class KM200 {

    private final KM200Device device;
    private final KM200Comm comm;
    private final ObjectMapper mapper = new ObjectMapper();
    private final HttpClient http;
    private final Duration timeout;
    private final String uri;
    private final FailsafeExecutor<HttpResponse<byte[]>> retry;

    /**
     * Configure the KM200 API.
     * 
     * This will also issue a silent query to /system to verify that you
     * configuration is correct.
     * 
     * Example:
     * 
     * <pre>
     * {@code
     * var uri = "http://192.168.0.44";
     * var gatewayPassword = "1234-5678-90ab-cdef";
     * var privatePassword = "secretExample";
     * var timeout = Duration.ofSeconds(5);
     * var salt = "1234567890aabbccddeeff11223344556677889900aabbccddeeffa0a1a2b2d3";
     * 
     * var km200 = new KM200(uri, timeout, gatewayPassword, privatePassword, salt);
     * 
     * // Read the heater's time
     * var time = km200.queryString("/gateway/DateTime");
     * System.out.println(time);
     * 
     * // Update the heater's time
     * km200.update("/gateway/DateTime", LocalDateTime.now());
     * 
     * // Explore the heater's endpoints
     * km200.endpoints().forEach(System.out::println);
     * }
     * </pre>
     * 
     * @param uri
     *            The base URI of your KM200 e.g. http://192.168.0.44
     * @param timeout
     *            An IO timeout for requests to your heater
     * @param gatewayPassword
     *            The constant gateway password which you need to read out from
     *            your heater's display e.g. 1234-5678-90ab-cdef
     * @param privatePassword
     *            The private password which you did assign in the app. If you
     *            forgot your private password you can start the "reset internet
     *            password" flow in the menu of your heater and then reassign a
     *            new passwort in the app.
     * @param salt
     *            The salt in the hexadecimal representation e.g. "12a0b2…"
     * 
     * @throws KM200Exception
     * @throws IOException
     * @throws InterruptedException
     */
    public KM200(String uri, Duration timeout, String gatewayPassword, String privatePassword, String salt)
            throws KM200Exception, IOException, InterruptedException {

        assertNotBlank(uri, "uri must not be blank");
        requireNonNull(timeout);
        assertNotBlank(gatewayPassword, "gatewayPassword must not be blank");
        assertNotBlank(privatePassword, "privatePassword must not be blank");
        assertNotBlank(salt, "salt must not be blank");

        var device = new KM200Device();
        device.setCharSet("UTF-8");
        device.setGatewayPassword(gatewayPassword.replace("-", ""));
        device.setPrivatePassword(privatePassword);
        device.setIP4Address(uri);
        device.setMD5Salt(salt);
        device.setInited(true);
        this.device = device;

        this.comm = new KM200Comm();
        this.timeout = timeout;
        this.uri = uri.replaceAll("/*$", "");
        this.http = newBuilder().connectTimeout(timeout).cookieHandler(new CookieManager()).followRedirects(ALWAYS)
                .build();

        retry = Failsafe
                .with(RetryPolicy.<HttpResponse<byte[]>> builder().handle(IOException.class).withMaxRetries(3).build());

        query("/system");
    }

    private final KM200Endpoint.Factory endpointFactory = new KM200Endpoint.Factory(this, mapper);

    public Stream<KM200Endpoint> endpoints() throws KM200Exception, IOException, InterruptedException {
        return endpointFactory.build();
    }

    private static record UpdateString(String value) {
    }

    public void update(String path, String value) throws KM200Exception, IOException, InterruptedException {
        var update = new UpdateString(value);
        update(path, update);
    }

    private static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss");

    public void update(String path, LocalDateTime time) throws KM200Exception, IOException, InterruptedException {
        update(path, time.format(DATE_TIME_FORMATTER));
    }

    private static record UpdateFloat(BigDecimal value) {
    }

    public void update(String path, int value) throws KM200Exception, IOException, InterruptedException {
        update(path, new BigDecimal(value));
    }

    public void update(String path, BigDecimal value) throws KM200Exception, IOException, InterruptedException {
        var update = new UpdateFloat(value);
        update(path, update);
    }

    private void update(String path, Object update) throws KM200Exception, IOException, InterruptedException {
        String json = null;
        try {
            json = mapper.writeValueAsString(update);
        } catch (JsonProcessingException e) {
            throw new IllegalStateException("Failed to update " + path, e);
        }
        var encrypted = comm.encodeMessage(device, json);
        if (encrypted == null) {
            throw new KM200Exception("Could not encrypt update " + json);
        }
        var request = request(path).POST(BodyPublishers.ofByteArray(encrypted)).build();
        var response = http.send(request, BodyHandlers.ofString());

        if (!(response.statusCode() >= 200 && response.statusCode() < 300)) {
            if (response.statusCode() == 423) {
                throw new KM200Exception.Locked(String.format("Failed to update %s with %s", path, json));
            }
            throw new KM200Exception(
                    String.format("Failed to update %s [%d]: %s", path, response.statusCode(), response.body()));
        }
    }

    public String query(String path) throws KM200Exception, IOException, InterruptedException {
        assertNotBlank(path, "Path must not be blank");
        if (!path.startsWith("/")) {
            throw new IllegalArgumentException("Path must start with a leading /");
        }

        var response = get(path);
        var encrypted = response.body();
        if (encrypted == null) {
            throw new KM200Exception("No response when querying " + path);
        }
        var decrypted = comm.decodeMessage(device, encrypted);
        if (decrypted == null) {
            throw new KM200Exception("Could not decrypt query " + path);
        }
        if (path.equals("/gateway/firmware")) {
            return decrypted;
        } else {
            if (!decrypted.startsWith("{")) {
                throw new KM200Exception(
                        String.format("Could not decrypt query %s. Body was:\n%s\n\n Decrypted was:\n%s", path,
                                encrypted, decrypted));
            }
        }
        return decrypted;
    }

    private HttpResponse<byte[]> get(String path) throws IOException, InterruptedException {
        var request = request(path).GET().build();

        HttpResponse<byte[]> response;
        try {
            response = retry.get(() -> http.send(request, BodyHandlers.ofByteArray()));
        } catch (FailsafeException e) {
            var cause = e.getCause();
            if (cause instanceof IOException) {
                throw (IOException) cause;

            } else if (cause instanceof InterruptedException) {
                throw (InterruptedException) cause;

            } else {
                throw e;
            }
        }

        return switch (response.statusCode()) {
        case 200 -> response;
        case 403 -> throw new KM200Exception.Forbidden("Query " + path + " is forbidden");
        case 404 -> throw new KM200Exception.NotFound("Query " + path + " was not found");
        default -> throw new KM200Exception("Query " + path + " failed with response code " + response.statusCode());
        };
    }

    public double queryDouble(String path) throws KM200Exception, IOException, InterruptedException {
        var json = queryJson(path);
        return json.get("value").asDouble();
    }

    public BigDecimal queryBigDecimal(String path) throws KM200Exception, IOException, InterruptedException {
        var json = queryJson(path);
        return json.get("value").decimalValue();
    }

    public String queryString(String path) throws KM200Exception, IOException, InterruptedException {
        var json = queryJson(path);
        return json.get("value").asText();
    }

    private JsonNode queryJson(String path) throws KM200Exception, IOException, InterruptedException {
        try {
            return mapper.readTree(query(path));
        } catch (JsonProcessingException e) {
            throw new KM200Exception("Could not parse JSON from query " + path, e);
        }
    }

    static final String USER_AGENT = "TeleHeater/2.2.3";

    private HttpRequest.Builder request(String path) {
        return HttpRequest.newBuilder(URI.create(uri + path)) //
                .setHeader("User-Agent", USER_AGENT) //
                .setHeader("Accept", "application/json") //
                .timeout(timeout);
    }

    private static void assertNotBlank(String var, String message) {
        if (requireNonNull(var).isBlank()) {
            throw new IllegalArgumentException(message);
        }
    }
}
